packages/flutter/lib/src/painting/debug_overflow_indicator.dart

```diff
@@ -0,0 +1,392 @@
+/// [overflowPaintIndicator] is surrounded by an assert).
+///
+/// Typical usage is to set the [overflowContainerRect] and the
+/// [overflowChildRect], and then call [overflowPaintIndicator]. This will
+/// calculate the overflow, and then draw the relevant indicators.
+///
+/// This class will also print a debug message to the console when the container
+/// overflows, but only on the first occurrence.  To reset and print it again
+/// upon the next occurrence, set [overflowReportNeeded] to true. This is
+/// typically done in the [RenderObject.reassemble] override in the
+/// [RenderObject] that uses the DebugOverflowIndicatorMixin.
+///
+/// ## Sample code
+///
+/// ```dart
+/// class MyRenderObject extends RenderObject with DebugOverflowContainerMixin {
+///   @override
+///   TextDirection get textDirection => _textDirection;
+///
+///   @override
+///   void performLayout() {
+///     // ...
+///     overflowChildRect = childParentData.offset & child.size;
+///     overflowContainerRect = Offset.zero & size;
+///   }
+///
+///   @override
+///   void paint(PaintingContext context, Offset offset) {
+///     if (isOverflowing) {
+///       _paintStuff(context, offset);
+///       return;
+///     }
+///     // ...
+///     assert(() {
+///       overflowRenderObject ??= this;
+///       overflowPaintIndicator(context, offset);
+///       return true;
+///     });
+///   }
+///
+///   @override
+///   void reassemble() {
+///     super.reassemble();
+///     assert(() {
+///       overflowReportNeeded = true;
+///       return true;
+///     }());
+///   }
+/// }
+/// ```
+///
+/// See also:
+///   * The code for [RenderUnconstrainedBox], and [RenderFlex] for examples of
+///     classes that use this indicator mixin.
+class DebugOverflowIndicatorMixin {
+  // This class is intended to be used as a mixin, and should not be
+  // extended directly.
+  factory DebugOverflowIndicatorMixin._() => null;
+
+  static const Color _black = const Color(0xBF000000);
+  static const Color _yellow = const Color(0xBFFFFF00);
+  static const double _kMarkerSize = 0.1;
+  static const double _debugMarkerFontSize = 7.5;
+  static const double _debugMarkerLabelPadding = 1.0;
+  static const TextStyle _debugMarkerTextStyle = const TextStyle(
+    color: const Color(0xFF900000),
+    fontSize: _debugMarkerFontSize,
+    fontWeight: FontWeight.w800,
+  );
+  static final Paint _debugMarkerPaint = new Paint()
+    ..shader = new ui.Gradient.linear(
+      const Offset(0.0, 0.0),
+      const Offset(10.0, 10.0),
+      <Color>[_black, _yellow, _yellow, _black],
+      <double>[0.25, 0.25, 0.75, 0.75],
+      TileMode.repeated,
+    );
+  static final Paint _debugLabelBackgroundPaint = new Paint()..color = const Color(0xFFFFFFFF);
+
+  final TextPainter _debugMarkerLabel = new TextPainter()
+    ..textDirection = TextDirection.ltr; // This label is in English.
+
+  TextDirection get textDirection => null;
+
+  RenderObject overflowRenderObject;
```

Comment:
I'd rather we didn't have this field. Why isn't it just an argument to overflowPaintIndicator()?

We can also make the mixin specifically a mixin at the rendering layer, in which case we can say it extends `RenderObject` and we can just use `this` everywhere instead of `overflowRenderObject`. As it is, this API only makes sense at the rendering layer, it's not truly reusable for other objects (since you require that it have a render object). We should either go all the way, or pull back and make it truly reusable.